### DISCUSSION TITLE
wt32-eth01 micropython 1.20 LAN Port not function

### DISCUSSION THREAD

**Discussion by winitboonkliang** [micropython/micropython#11446] [ESP32]
wt32-eth01 (V1.4 PCB)  Lan Error 

import network
import time
import machine
from machine import UART
from machine import Pin

lan = network.LAN(mdc = machine.Pin(23), mdio = machine.Pin(18), power = None, phy_type = network.PHY_LAN8720, phy_addr=0)
lan.active(1)
print('ifconfig :', lan.ifconfig())
print('status : ', lan.status())
print('isconnected : ', lan.isconnected())
print('active : ', lan.active())

OSError: esp_eth_set_default_handlers failed (invalid parameter)

**Comment by elvis-epx** [comment]
I actually managed to make it work:

import network
import machine
lan = network.LAN(mdc=machine.Pin(23), mdio=machine.Pin(18), phy_type=network.PHY_LAN8720, phy_addr=1, power=machine.Pin(16), id=0)
lan.active(true)
lan.ifconfig()

Two crucial things: 
a) the power pin is 16,
b) you must not have tried to create another network.LAN (the error changes if you try it twice with wrong parameters, so it gives off that it is not stateless). I think it keeps state even across soft reboots. Always disconnect power, reconnect, and start fresh.

**Reply by dpgeorge** [reply]
Docs updated in cddeb5f0754e42f5e50da8ccefb5c073dce61be6, thanks!

**Comment by cobradevil** [comment]
So this is almost the same issue but different, didn't know for sure to open a new discussion or to revive this one.

I have installed 1.25 and also build the newest 1.26-master (using idf-5.4.1) but i have issues with this board.
I think it should be the same as the one above (cheap clone aliexpress) but with the same settings i get the following:

```
>>> import machine,network                                                                                                                                                              
>>> lan = network.LAN(mdc=machine.Pin(23), mdio=machine.Pin(18), phy_type=network.PHY_LAN8720, phy_add)
E (10114) eth_phy_802_3: esp_eth_phy_802_3_pwrctl(244): power up timeout                               
E (10114) eth_phy_802_3: esp_eth_phy_802_3_basic_phy_init(413): power control failed                   
E (10114) lan87xx: lan87xx_init(326): failed to init PHY                                               
E (10124) esp_eth: esp_eth_driver_install(229): init phy failed
```

When I also add the options ref_clk and ref_clk_mode then it gets initialized:
```
>>> import machine,network                                                                             
>>> lan = network.LAN(                                                                                 
...     mdc=machine.Pin(23),                                                                           
...     mdio=machine.Pin(18),                                                                          
...     power=machine.Pin(16),                                                                         
...     phy_type=network.PHY_LAN8720,                                                                  
...     phy_addr=1,                                                                                    
...     ref_clk=machine.Pin(0),                                                                        
...     ref_clk_mode=machine.Pin.OUT                                                                   
... )                                                                                                  
>>> lan.active(1)                                                                                      
True
```

But using: ref_clk_mode=machine.Pin.IN

```
>>> import machine,network                                                      
>>> lan = network.LAN(                                                          
...     mdc=machine.Pin(23),                                                    
...     mdio=machine.Pin(18),                                                   
...     power=machine.Pin(16),                                                  
...     phy_type=network.PHY_LAN8720,                                           
...     phy_addr=1,                                                             
...     ref_clk=machine.Pin(0),                                                 
...     ref_clk_mode=machine.Pin.IN)                                            
E (22314) eth_phy_802_3: esp_eth_phy_802_3_pwrctl(244): power up timeout        
E (22314) eth_phy_802_3: esp_eth_phy_802_3_basic_phy_init(413): power control fd
E (22314) lan87xx: lan87xx_init(326): failed to init PHY                        
E (22324) esp_eth: esp_eth_driver_install(229): init phy failed 
```

The thing is that when it gets initialized correctly then i have no ip connectivity accept when i ground io0, then i can see dhcp messages coming in using tcpdump.

When using the same board with tasmota it works like a charm, when loading an arduino sketch it works fine so why does it not work with micropython? I have been breaking my head over this one for more then 2 days.

```
#include <ETH.h>

IPAddress localIP(192, 168, 1, 50);
IPAddress gateway(192, 168, 1, 1);
IPAddress subnet(255, 255, 255, 0);
IPAddress dns1(8, 8, 8, 8);
IPAddress dns2(1, 1, 1, 1);

#define ETH_PHY_TYPE     ETH_PHY_LAN8720
#define ETH_PHY_ADDR     1
#define ETH_MDC_PIN      23
#define ETH_MDIO_PIN     18
#define ETH_POWER_PIN    16
#define ETH_CLK_MODE     ETH_CLOCK_GPIO0_OUT // of ETH_CLOCK_GPIO17_OUT voor RMII out via GPIO17

void setup() {
  Serial.begin(115200);
  delay(1000);
  Serial.println("Ethernet Init");

  // Start Ethernet
  ETH.begin(
    ETH_PHY_TYPE,
    ETH_PHY_ADDR,
    ETH_MDC_PIN,
    ETH_MDIO_PIN,
    ETH_POWER_PIN,
    ETH_CLK_MODE
  );
  // Static IP config BEFORE begin()
  ETH.config(localIP, gateway, subnet, dns1, dns2);
}

void loop() {
  Serial.print("IP address: ");
  Serial.println(ETH.localIP());
  delay(5000);
}
```

**Reply by robert-hh** [reply]
First a note from README.md: `Currently MicroPython supports v5.2, v5.2.2, v5.3 and v5.4`. So you might use one of the supported idf versions. 
Besides, setting ref_clk_mode to OUT may be the right setting, if lan.active(1) returns True. As noted above, you have to run a hard reset between attempts.
Since lan.active(1) returns True, the mdc connection is fine and as well the mdc address.

Cross check for ref_clk_mode:
If I try to see a schematic of this board, it seems that the LAN8720 has it's own 50MHz clock, and so the ref_clk_mode should be IN. Does your board have a 50MHz crystal?. It is a small rectangular metal package with the number 50.0 printed on it.

**Reply by cobradevil** [reply]
Yes i saw that from the readme about the idf versions but i also used the original bin 1.25 version to no avail.

So yes it has a 50mhz clock so i think there is something wrong then using the init code. I always do a hard reset by pulling the power. So in the above third attempt you can see i use the machine.Pin.IN but it does not initialize ok. only when i set the pin.OUT it does initialize ok but then i have to ground GPIO0 to have at least a working ping to the device. So what could i try? Not really a programmer here but chatgpt does help a bit :). I also tried the sketch to use GPIO0 IN but then i also get a error: 

```
Ethernet Init                                                                                          
E (1118) eth_phy_802_3: esp_eth_phy_802_3_pwrctl(307): power up timeout                                
E (1118) eth_phy_802_3: esp_eth_phy_802_3_basic_phy_init(510): power control failed                    
E (1121) lan87xx: lan87xx_init(339): failed to init PHY                                                
E (1125) esp_eth: esp_eth_driver_install(229): init phy failed                                         
IP address: 0.0.0.0                                                 
```

**Reply by robert-hh** [reply]
If the device responds to Ping, then the TCP/IP stack is working to some extend. You can try to omit the power keyword. The power pin controls the 50Mhz oscillator. If not set, the 50Mhz crystal is not used. Recently, that worked for a user.

**Reply by cobradevil** [reply]
Hi Robert,
as you can see i also did that in the examples in my first post. I can only ping the device when grounding GPIO0 and i do not have enough knowledge to know what that means. By accident i saw the device coming alive when using Pin.OUT and grounding GPIO0 to go into flashing mode.

**Reply by robert-hh** [reply]
You did not specify ref_clk and ref_clk_mode in the first post.

**Reply by robert-hh** [reply]
I do not know if you managed to get LAN working meanwhile. I got a WT32 board today. LAN is working with the following start-up code:
```
import machine, network
lan=network.LAN(mdc=machine.Pin(23), mdio=machine.Pin(18), power=machine.Pin(16),
    phy_type=network.PHY_LAN8720, phy_addr=1,
    ref_clk_mode=machine.Pin.IN, ref_clk=machine.Pin(0))
lan.active(True)
```
I can run test code like iperf3.py successfully.

**Reply by shariltumin** [reply]
Thanks! I also manage to connect to my local ethernet.

```py
# setup_eth.py
#
def get_lan():
    lan=network.LAN(mdc=machine.Pin(23), mdio=machine.Pin(18), power=machine.Pin(16),
    phy_type=network.PHY_LAN8720, phy_addr=1,
    ref_clk_mode=machine.Pin.IN, ref_clk=machine.Pin(0))
    w = lan.active(True)
    print('LAN active/status:', w, lan.status())
    ip = lan.ifconfig()
    print('LAN IP:', ip)
```

When the board is correctly connected to ethernet switch:

```
>>> from setup_eth import get_lan
>>> get_lan()
LAN active/status: True 5
LAN IP: ('192.168.5.20', '255.255.252.0', '192.168.4.1', '192.168.4.1')
```

And when it is not connected:
```
>>> from setup_eth import get_lan
>>> get_lan()
LAN active/status: True 4
LAN IP: ('0.0.0.0', '0.0.0.0', '0.0.0.0', '192.168.4.1')
```
